fix #45361, fix #281083: remove extra annotations and spanners on paste and note duration change

dmitrio95 opened the pull request:
This change is intended to fix [this issue](https://musescore.org/en/node/45361) as well as the related copy-paste issues mentioned in [the recent video](https://musescore.org/en/handbook/developers-handbook/ux-design/design-reviews-and-responses/tantacrul-music-software#28%3A53_-_Copy_paste_with_overwriting_of_markings) by Tantacrul.

This fix makes paste operation behave like if the user has selected the target range (with the actual selection filter settings), deleted its contents and pasted after that the copied range. This
eliminates the problem of having some of the previous elements in the target range.

Although this behavior sounds logical, there may be some possible situations I could have missed when considering this solution. Also I made the paste operation delete elements according to the *current* selection filter settings rather than those that were applied when the selection was initially copied. This should work in most common cases but maybe recording the selection filter state to the pasted data can make sense, I am not sure which behavior will be less expected by a user in case the current and the copied selection filter settings are different.

Anyway, while this patch should work well in most situations, it would still be better to test it better in some corner cases or, maybe, in some real score editing scenarios.

MarcSabatella commented:
What I remember about the history of this, is there were all sorts of specific cases like with hairpins that started before the pasted area but ended within it, or voltas that needed to remain, etc, and there about six or seven different attempts at handling this before the current scheme was settled on.  Each time some change was made, the next day there would be complaints that while it fixed one case it broke another.  This was some point during 2.0 development, maybe early in the beta phase.  Not that helpful I know, but anyhow, I do think you're right to think that we need to tread carefully here.

That said, I do like the overall approach because it makes it consistent with delete.  Probably if people don't like the paste behavior they wouldn't like the delete behavior either, and yet we don't get complaints about that so much.

dmitrio95 commented:
As regression tests showed me that I was a bit wrong about my solution, here is an alternative approach.
The elements are pasted to the gaps that are specially prepared by `makeGap1()` function. As these gaps perfectly match what we actually need to clean up from extra elements, I moved the code that did the deletion of extra elements to `makeGap1()`. Also I limited it to removing only spanners and annotations (dynamic markings etc.) as chords and rests are already handled by this function properly.

Also I added the commit that uses a similar approach to fix a [similar issue](https://musescore.org/en/node/281083) on note duration change. Except for some features of `makeGap()` vs. `makeGap1()` functions the solution is absolutely the same.

Hope this will work better.

dmitrio95 commented:
The test failure is exactly due to removing annotations on note duration extending: chord symbols also get deleted while the test checks that chord symbols are present. Given that chord symbols can exist without note or rests retaining them can make sense for them. Maybe annotations removal should be limited to any annotations except chord symbols or even only dynamic markings — with this patch this is a matter of setting a proper `filter` argument for `removeAnnotationsFromRange`.

dmitrio95 commented:
Excluded chord symbols from the list of deleted elements on operations using `makeGap` (like note duration change), changed a bit the spanners removal condition to match the condition of their range selection.

MarcSabatella commented:
Thanks, I agree on leaving chord symbols, it's definitely expected behavior in most cases.  Right now on paste I believe we do some special casing to replace a chord symbol at the same tick (both to avoid duplication and to pick up any changes) but keep existing ones.

MarcSabatella commented:
Right now chord symbols are *not* being preserved, so somehow this didn't work, or there is some misunderstanding here.
